File: microservices/user-service/app/constants/tags.py
# microservices/user-service/app/constants/tags.py
from enum import Enum


class AgeTag(str, Enum):
    """Tags for age groups."""
    CHILDREN = "0100"  # Trẻ em
    TEENAGER = "0101"  # Vị thành niên
    ADULT = "0102"  # Người trưởng thành
    MIDDLE_AGED = "0103"  # Trung niên
    ELDER = "0104"  # Cao tuổi


class MedicalConditionTag(str, Enum):
    """Tags for medical conditions."""

    # Cardiovascular - Bệnh tim mạch
    CARDIOVASCULAR_DISEASE = "0200"  # Bệnh tim mạch
    HIGH_BLOOD_PRESSURE = "0201"  # Huyết áp cao
    LOW_BLOOD_PRESSURE = "0202"  # Huyết áp thấp
    BLOOD_CLOTTING_DISORDER = "0203"  # Rối loạn đông máu

    # Metabolic - Bệnh chuyển hóa
    METABOLIC_DISEASE = "0210"  # Bệnh chuyển hóa
    OBESITY = "0211"  # Béo phì
    DIABETES = "0212"  # Tiểu đường
    GOUT = "0213"  # Gout

    # Liver & Gallbladder - Gan & Túi mật
    FATTY_LIVER = "0220"  # Gan nhiễm mỡ
    LIVER_FAILURE = "0221"  # Suy gan
    CIRRHOSIS = "0222"  # Xơ gan
    GALLSTONES = "0223"  # Sỏi mật

    # Kidney - Thận
    KIDNEY_FAILURE = "0230"  # Suy thận
    KIDNEY_STONES = "0231"  # Sỏi thận

    # Bone & Joint - Xương & Khớp
    OSTEOPOROSIS = "0240"  # Loãng xương
    RICKETS = "0241"  # Còi xương
    ARTHRITIS = "0242"  # Viêm khớp

    # Digestive - Tiêu hóa
    GASTRIC_ULCER = "0250"  # Viêm loét dạ dày
    ACID_REFLUX = "0251"  # Trào ngược dạ dày
    DIARRHEA = "0252"  # Tiêu chảy
    CONSTIPATION = "0253"  # Táo bón
    TOOTH_DECAY = "0254"  # Sâu răng

    # Nutritional - Dinh dưỡng
    MALNUTRITION = "0260"  # Suy dinh dưỡng
    ANEMIA = "0261"  # Thiếu máu
    POOR_EYESIGHT = "0262"  # Mắt kém
    HAIR_LOSS = "0263"  # Rụng tóc

    # Other - Khác
    INFLAMMATION = "0270"  # Viêm mạn tính
    WEAK_IMMUNITY = "0271"  # Miễn dịch yếu
    DERMATOLOGICAL_DISEASE = "0272"  # Bệnh da liễu


class AllergyTag(str, Enum):
    """Tags for food allergies."""
    SHELLFISH = "0300"  # Hải sản có vỏ
    FISH = "0301"  # Cá
    DAIRY = "0302"  # Sữa
    EGG = "0303"  # Trứng
    PEANUT = "0304"  # Đậu phộng
    SOY = "0305"  # Đậu nành
    SESAME = "0306"  # Mè
    NUTS = "0307"  # Các loại hạt
    WHEAT = "0308"  # Lúa mì


class SpecialDietTag(str, Enum):
    """Tags for special dietary preferences."""
    GYMER = "0400"  # Người tập gym
    DIETER = "0401"  # Người ăn kiêng
    VEGAN = "0402"  # Người ăn chay trường (thuần chay)
    VEGETARIAN = "0403"  # Người ăn chay (có thể dùng sữa, trứng)


class TastePreferenceTag(str, Enum):
    """Tags for taste preferences."""
    SALTY_PREF = "0500"  # Thích mặn
    SWEET_PREF = "0501"  # Thích ngọt
    UMAMI_PREF = "0502"  # Thích umami
    SOUR_PREF = "0503"  # Thích chua
    BITTER_PREF = "0504"  # Thích đắng
    SPICY_PREF = "0505"  # Thích cay
    LIGHT_PREF = "0506"  # Thích thanh đạm
    RICH_PREF = "0507"  # Thích béo

File: webapp/src/services/tags.ts
// Tag types and constants based on backend user-service/app/enums/tags.py

export type TagCategory = 'age' | 'medical' | 'allergy' | 'diet' | 'taste';

export type TagValue =
  // Age tags
  | '0100' | '0101' | '0102' | '0103' | '0104'
  // Medical condition tags
  | '0200' | '0201' | '0202' | '0203'
  | '0210' | '0211' | '0212' | '0213'
  | '0220' | '0221' | '0222' | '0223'
  | '0230' | '0231'
  | '0240' | '0241' | '0242'
  | '0250' | '0251' | '0252' | '0253' | '0254'
  | '0260' | '0261' | '0262' | '0263'
  | '0270' | '0271' | '0272'
  // Allergy tags
  | '0300' | '0301' | '0302' | '0303' | '0304' | '0305' | '0306' | '0307' | '0308'
  // Diet tags
  | '0400' | '0401' | '0402' | '0403'
  // Taste preference tags
  | '0500' | '0501' | '0502' | '0503' | '0504' | '0505' | '0506' | '0507';

export interface UserTag {
  id: number;
  tag_value: TagValue;
  tag_category: TagCategory;
  tag_name: string;
  description: string;
  updated_at: string;
}

export interface UserTagsData {
  age: UserTag[];
  medical: UserTag[];
  allergy: UserTag[];
  diet: UserTag[];
  taste: UserTag[];
}

export interface UserTagsResponse {
  data: UserTagsData;
  total_tags: number;
  categories_count: {
    age: number;
    medical: number;
    allergy: number;
    diet: number;
    taste: number;
  };
}

// Age tags
export const AGE_TAGS = {
  CHILDREN: { value: '0100' as TagValue, name: 'Trẻ em', nameEn: 'CHILDREN' },
  TEENAGER: { value: '0101' as TagValue, name: 'Vị thành niên', nameEn: 'TEENAGER' },
  ADULT: { value: '0102' as TagValue, name: 'Người trưởng thành', nameEn: 'ADULT' },
  MIDDLE_AGED: { value: '0103' as TagValue, name: 'Trung niên', nameEn: 'MIDDLE_AGED' },
  ELDER: { value: '0104' as TagValue, name: 'Cao tuổi', nameEn: 'ELDER' },
} as const;

// Medical condition tags
export const MEDICAL_TAGS = {
  // Cardiovascular
  CARDIOVASCULAR_DISEASE: { value: '0200' as TagValue, name: 'Bệnh tim mạch', nameEn: 'CARDIOVASCULAR_DISEASE' },
  HIGH_BLOOD_PRESSURE: { value: '0201' as TagValue, name: 'Huyết áp cao', nameEn: 'HIGH_BLOOD_PRESSURE' },
  LOW_BLOOD_PRESSURE: { value: '0202' as TagValue, name: 'Huyết áp thấp', nameEn: 'LOW_BLOOD_PRESSURE' },
  BLOOD_CLOTTING_DISORDER: { value: '0203' as TagValue, name: 'Rối loạn đông máu', nameEn: 'BLOOD_CLOTTING_DISORDER' },
  // Metabolic
  METABOLIC_DISEASE: { value: '0210' as TagValue, name: 'Bệnh chuyển hóa', nameEn: 'METABOLIC_DISEASE' },
  OBESITY: { value: '0211' as TagValue, name: 'Béo phì', nameEn: 'OBESITY' },
  DIABETES: { value: '0212' as TagValue, name: 'Tiểu đường', nameEn: 'DIABETES' },
  GOUT: { value: '0213' as TagValue, name: 'Gout', nameEn: 'GOUT' },
  // Liver & Gallbladder
  FATTY_LIVER: { value: '0220' as TagValue, name: 'Gan nhiễm mỡ', nameEn: 'FATTY_LIVER' },
  LIVER_FAILURE: { value: '0221' as TagValue, name: 'Suy gan', nameEn: 'LIVER_FAILURE' },
  CIRRHOSIS: { value: '0222' as TagValue, name: 'Xơ gan', nameEn: 'CIRRHOSIS' },
  GALLSTONES: { value: '0223' as TagValue, name: 'Sỏi mật', nameEn: 'GALLSTONES' },
  // Kidney
  KIDNEY_FAILURE: { value: '0230' as TagValue, name: 'Suy thận', nameEn: 'KIDNEY_FAILURE' },
  KIDNEY_STONES: { value: '0231' as TagValue, name: 'Sỏi thận', nameEn: 'KIDNEY_STONES' },
  // Bone & Joint
  OSTEOPOROSIS: { value: '0240' as TagValue, name: 'Loãng xương', nameEn: 'OSTEOPOROSIS' },
  RICKETS: { value: '0241' as TagValue, name: 'Còi xương', nameEn: 'RICKETS' },
  ARTHRITIS: { value: '0242' as TagValue, name: 'Viêm khớp', nameEn: 'ARTHRITIS' },
  // Digestive
  GASTRIC_ULCER: { value: '0250' as TagValue, name: 'Viêm loét dạ dày', nameEn: 'GASTRIC_ULCER' },
  ACID_REFLUX: { value: '0251' as TagValue, name: 'Trào ngược dạ dày', nameEn: 'ACID_REFLUX' },
  DIARRHEA: { value: '0252' as TagValue, name: 'Tiêu chảy', nameEn: 'DIARRHEA' },
  CONSTIPATION: { value: '0253' as TagValue, name: 'Táo bón', nameEn: 'CONSTIPATION' },
  TOOTH_DECAY: { value: '0254' as TagValue, name: 'Sâu răng', nameEn: 'TOOTH_DECAY' },
  // Nutritional
  MALNUTRITION: { value: '0260' as TagValue, name: 'Suy dinh dưỡng', nameEn: 'MALNUTRITION' },
  ANEMIA: { value: '0261' as TagValue, name: 'Thiếu máu', nameEn: 'ANEMIA' },
  POOR_EYESIGHT: { value: '0262' as TagValue, name: 'Mắt kém', nameEn: 'POOR_EYESIGHT' },
  HAIR_LOSS: { value: '0263' as TagValue, name: 'Rụng tóc', nameEn: 'HAIR_LOSS' },
  // Other
  INFLAMMATION: { value: '0270' as TagValue, name: 'Viêm mạn tính', nameEn: 'INFLAMMATION' },
  WEAK_IMMUNITY: { value: '0271' as TagValue, name: 'Miễn dịch yếu', nameEn: 'WEAK_IMMUNITY' },
  DERMATOLOGICAL_DISEASE: { value: '0272' as TagValue, name: 'Bệnh da liễu', nameEn: 'DERMATOLOGICAL_DISEASE' },
} as const;

// Allergy tags
export const ALLERGY_TAGS = {
  SHELLFISH: { value: '0300' as TagValue, name: 'Hải sản có vỏ', nameEn: 'SHELLFISH' },
  FISH: { value: '0301' as TagValue, name: 'Cá', nameEn: 'FISH' },
  DAIRY: { value: '0302' as TagValue, name: 'Sữa', nameEn: 'DAIRY' },
  EGG: { value: '0303' as TagValue, name: 'Trứng', nameEn: 'EGG' },
  PEANUT: { value: '0304' as TagValue, name: 'Đậu phộng', nameEn: 'PEANUT' },
  SOY: { value: '0305' as TagValue, name: 'Đậu nành', nameEn: 'SOY' },
  SESAME: { value: '0306' as TagValue, name: 'Mè', nameEn: 'SESAME' },
  NUTS: { value: '0307' as TagValue, name: 'Các loại hạt', nameEn: 'NUTS' },
  WHEAT: { value: '0308' as TagValue, name: 'Lúa mì', nameEn: 'WHEAT' },
} as const;

// Diet tags
export const DIET_TAGS = {
  GYMER: { value: '0400' as TagValue, name: 'Người tập gym', nameEn: 'GYMER' },
  DIETER: { value: '0401' as TagValue, name: 'Người ăn kiêng', nameEn: 'DIETER' },
  VEGAN: { value: '0402' as TagValue, name: 'Người ăn chay trường', nameEn: 'VEGAN' },
  VEGETARIAN: { value: '0403' as TagValue, name: 'Người ăn chay', nameEn: 'VEGETARIAN' },
} as const;

// Taste preference tags
export const TASTE_TAGS = {
  SALTY_PREF: { value: '0500' as TagValue, name: 'Thích mặn', nameEn: 'SALTY_PREF' },
  SWEET_PREF: { value: '0501' as TagValue, name: 'Thích ngọt', nameEn: 'SWEET_PREF' },
  UMAMI_PREF: { value: '0502' as TagValue, name: 'Thích umami', nameEn: 'UMAMI_PREF' },
  SOUR_PREF: { value: '0503' as TagValue, name: 'Thích chua', nameEn: 'SOUR_PREF' },
  BITTER_PREF: { value: '0504' as TagValue, name: 'Thích đắng', nameEn: 'BITTER_PREF' },
  SPICY_PREF: { value: '0505' as TagValue, name: 'Thích cay', nameEn: 'SPICY_PREF' },
  LIGHT_PREF: { value: '0506' as TagValue, name: 'Thích thanh đạm', nameEn: 'LIGHT_PREF' },
  RICH_PREF: { value: '0507' as TagValue, name: 'Thích béo', nameEn: 'RICH_PREF' },
} as const;

// Tag category display names
export const TAG_CATEGORY_NAMES: Record<TagCategory, string> = {
  age: 'Độ tuổi',
  medical: 'Bệnh lý',
  allergy: 'Dị ứng',
  diet: 'Chế độ ăn',
  taste: 'Khẩu vị',
};

// Get tag by value
export function getTagByValue(value: TagValue) {
  const allTags = {
    ...AGE_TAGS,
    ...MEDICAL_TAGS,
    ...ALLERGY_TAGS,
    ...DIET_TAGS,
    ...TASTE_TAGS,
  };

  for (const tag of Object.values(allTags)) {
    if (tag.value === value) {
      return tag;
    }
  }
  return null;
}
